### ISSUE TITLE
netkvm: does it need to wait when StopFlows

### ISSUE THREAD

**Issue by wangyan0507** [virtio-win/kvm-guest-drivers-windows#1221]
When shutdown the windows vm with netkvm driver, shoud it wait the m_Counter to zero for m_RxStateMachine?
The shutdown stack:
```
ndis!ndisDispatchIoWorkItem
->netkvm!OnSetPowerWorkItem
  ->netkvm!ParaNdis_PowerOff
    ->netkvm!CMiniportStateMachine::NotifySuspended
      ->netkvm!CMiniportStateMachine::StopFlows
        ->netkvm!CNdisList<CDataFlowStateMachine,CRawAccess,CNonCountingObject>::ForEach
          ->netkvm!CFlowStateMachine::Stop
            ->m_NoOutstandingItems.Wait();  // which will wait the m_Counter to become StoppedMask.
```
For shutdown vm scene, when there is a lot of incomping packets, the m_Counter will be released as follows for m_RxStateMachine:
```
ParaNdis6_ReturnNetBufferLists()
  ->pContext->m_RxStateMachine.UnregisterOutstandingItems(nofNetkvm);  // which will release for m_Counter.
    ->m_Counter.Release(NumItems)
    -> CheckCompletion(value)
      ->CompleteStopping()
        ->m_NoOutstandingItems.Notify(); // which will awake the upper 'm_NoOutstandingItems.Wait()'.
```
But from the microsoft doc for [OID_PNP_SET_POWER](https://learn.microsoft.com/en-us/windows-hardware/drivers/network/oid-pnp-set-power#transitioning-to-a-low-power-state-d1-d3):
The miniport driver that supports NDIS 6.30 and later versions of NDIS must also do the following:
Not wait for the completion of pending receive indications through calls to its [MiniportReturnNetBufferLists](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ndis/nc-ndis-miniport_return_net_buffer_lists) function. 

Thanks in advance!
Yan Wang

**Comment by YanVugenfirer** [comment]
@wangyan0507 Unfortunately, paravirtualizaed "HW" behaves differently than actual HW in this case, so we must wait.
If we release the memory (the NET_BUFFER_LISTs), the host might still continue to write to memory that can potentially be reused by the guest.

**Comment by wangyan0507** [comment]
@YanVugenfirer Sorry, I am not understand very well. 
The shutdown process:
```
ndis!ndisDispatchIoWorkItem
->netkvm!OnSetPowerWorkItem
  ->netkvm!ParaNdis_PowerOff
    ->netkvm!CMiniportStateMachine::NotifySuspended
      ->netkvm!CMiniportStateMachine::StopFlows
        ->netkvm!CNdisList<CDataFlowStateMachine,CRawAccess,CNonCountingObject>::ForEach
          ->netkvm!CFlowStateMachine::Stop
            ->m_Counter.SetMask(StoppedMask); // Or with StoppedMask to m_Counter.
            ->m_NoOutstandingItems.Wait();    // Wait the m_Counter to become StoppedMask.
```
The normal rx process:
```
ndis!ndisMiniportDpc
  ->netkvm!MiniportMSIInterruptDpc
    ->netkvm!ParaNdis_RXTXDPCWorkBody
      ->netkvm!RxDPCWorkBody
        ->pContext->m_RxStateMachine.RegisterOutstandingItems // Add reference for m_Counter.
        ->ndis!NdisMIndicateReceiveNetBufferLists
          ->...
            ->netkvm!ParaNdis6_ReturnNetBufferLists
              ->pContext->m_RxStateMachine.UnregisterOutstandingItems
                ->m_Counter.Release(NumItems) // Release NumItems for m_Counter.
                -> CheckCompletion(value)
                  ->CompleteStopping()
                    ->m_NoOutstandingItems.Notify(); // Wake the upper 'm_NoOutstandingItems.Wait()'.
```
In "Not wait for the completion of pending receive indications through calls to its [MiniportReturnNetBufferLists](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ndis/nc-ndis-miniport_return_net_buffer_lists) function.", does it mean no need to wait the call for netkvm!ParaNdis6_ReturnNetBufferLists? If it is, the call m_NoOutstandingItems.Wait() may be get stuck, because it depends on the m_Counter to become StoppedMask.
You mentioned "If we release the memory (the NET_BUFFER_LISTs)", I don't know get it why we release the memory?

Thanks!
Yan Wang

**Comment by YanVugenfirer** [comment]
@wangyan0507 Do you see an actual case when we are stuck waiting for completed buffers, or is it a theoretical question?

**Comment by wangyan0507** [comment]
@YanVugenfirer Yes, I got a stuck waiting for completed buffers in the older netkvm drivers. The bosd RIVER_POWER_STATE_FAILURE (9f) happend.
The stuck stack is as follows, the m_Counter is 1 with the StoppedMask.
![image](https://github.com/user-attachments/assets/b35a779b-485d-4046-a865-f21b95af4cc5)

**Comment by YanVugenfirer** [comment]
Can you please send more details, like what OS, what drivers version, and the exact scenario?

Thank you.

**Comment by ybendito** [comment]
@wangyan0507 Please let us know the exact OS version where the problem happens. We need to reproduce this behavior before continuing the discussion

**Comment by wangyan0507** [comment]
@YanVugenfirer @ybendito  I am very sorry for forgetting to reply.

***Virtual Machine***: Windows 11 24H2 arm64, open the "fast startup" option.
***NetKvm Driver Version***: using the tag mm249 to compile the NetKvm Driver.
***Method of reproduction***:
1) using "ping VM_IP -i 0.01 -c 100000000" to ping the windows vm all the time.
2) repeating "shutdown" and "start up" windows virtual maching, the interval is 30-60 seconds. About 30-60 minutes, it will cause the the bsod DRIVER_POWER_STATE_FAILURE (9f).

**Comment by ybendito** [comment]
Can you please try whether this one is better (based on tag 286)
https://github.com/virtio-win/kvm-guest-drivers-windows/tree/issue1221

**Comment by wangyan0507** [comment]
@ybendito 
I use the tag mm249 with your patch "potention fix for testing only" to compile the NetKvm Driver.  Currently, the method of reproduction has been run for 5 hours, not cause bosd. I think it works.

Thanks,
Yan Wang

**Comment by ybendito** [comment]
https://issues.redhat.com/browse/RHEL-135728

**Comment by YanVugenfirer** [comment]
This issue has been tracked in Jira: https://redhat.atlassian.net/browse/RHELMISC-34260
